### PR TITLE
server: stop listener on critical error.

### PR DESCRIPTION
**Pull request by coocood**
When binlog failed to write, every write is going to fail.
We should stop serving to let clients use other TiDB servers.

Mannually tested with pump.

 Kill the pump, then clients lost connection and can not be connected again.

### CODE REVIEW

**Comment by coocood** [comment]
@GregoryIan @shenli PTAL

File: server/server.go [hunk 1]
```diff
@@ -171,8 +177,21 @@ func (s *Server) Run() error {
...
+		case <-s.stopListenerCh:
+			conn.Close()
+			break out
+		default:
+		}
+
 		go s.onConn(conn)
 	}
+	s.listener.Close()
+	s.listener = nil
```
Comment (shenli) [inline server/server.go]: Should we add a metrics and alert for this condition?

File: server/server.go [hunk 2]
```diff
@@ -70,6 +70,11 @@ type Server struct {
 	rwlock            *sync.RWMutex
 	concurrentLimiter *TokenLimiter
 	clients           map[uint32]*clientConn
+
+	// When a critical error occurred, we don't want to exit the process, because there may be
+	// a supervisor automatically restart it, then new client connection will be created, but we can't server it.
+	// So we just stop the listener and store to force clients to chose other TiDB servers.
+	stopListenerCh chan struct{}
```
Comment (shenli) [inline server/server.go]: In some users' scenario, they do not use lb to access tidb-server.

File: sessionctx/binloginfo/binloginfo.go [hunk 3]
```diff
@@ -46,11 +49,21 @@ func GetPrewriteValue(ctx context.Context, createIfNotExists bool) *binlog.Prewr
...
 	commitData, _ := bin.Marshal()
 	req := &binlog.WriteBinlogReq{ClusterID: clusterID, Payload: commitData}
-	resp, err := PumpClient.WriteBinlog(goctx.Background(), req)
-	if err == nil && resp.Errmsg != "" {
-		err = errors.New(resp.Errmsg)
+
+	// Retry 3 times because we may raise CRITICAL error here.
+	var err error
+	for i := 0; i < 3; i++ {
+		var resp *binlog.WriteBinlogResp
```
Comment (shenli) [inline sessionctx/binloginfo/binloginfo.go]: Should we add a metrics here?

File: server/server.go (hunk 2 above)
Comment (GregoryIan) [inline server/server.go]: In those scenarios, this tidb-server is still unavailable. I don't think it's a problem if we add an alert for it.

File: sessionctx/binloginfo/binloginfo.go [hunk 4]
```diff
@@ -46,11 +49,21 @@ func GetPrewriteValue(ctx context.Context, createIfNotExists bool) *binlog.Prewr
...
+		resp, err = PumpClient.WriteBinlog(goctx.Background(), req)
+		if err == nil && resp.Errmsg != "" {
+			err = errors.New(resp.Errmsg)
+		}
+		if err == nil {
+			return nil
+		}
+		time.Sleep(300 * time.Millisecond)
```
Comment (GregoryIan) [inline sessionctx/binloginfo/binloginfo.go]: it's too short for some recoverable errors . how long we can tolerate service is unavailable? is it better to add it in arguments?

File: terror/terror.go [hunk 5]
```diff
@@ -39,10 +35,8 @@ type ErrCode int
 
 // Executor error codes.
 const (
-	CodeUnknown                  ErrCode = -1
-	CodeCommitNotInTransaction           = 1
-	CodeRollbackNotInTransaction         = 2
-	CodeExecResultIsEmpty                = 3
+	CodeUnknown           ErrCode = -1
+	CodeExecResultIsEmpty         = 3
```
Comment (GregoryIan) [inline terror/terror.go]: `CodeCommitNotInTransaction` is useless? And Is other errors use CodeExecResultIsEmpty?  If not, set it to 1

Comment (coocood) [inline terror/terror.go]: No, it's not used.

**Comment by zimulala** [comment]
LGTM

File: sessionctx/binloginfo/binloginfo.go [hunk 6]
```diff
@@ -46,11 +49,21 @@ func GetPrewriteValue(ctx context.Context, createIfNotExists bool) *binlog.Prewr
...
-	if err == nil && resp.Errmsg != "" {
-		err = errors.New(resp.Errmsg)
+
+	// Retry many times because we may raise CRITICAL error here.
+	var err error
+	for i := 0; i < 20; i++ {
+		var resp *binlog.WriteBinlogResp
+		resp, err = PumpClient.WriteBinlog(goctx.Background(), req)
+		if err == nil && resp.Errmsg != "" {
+			err = errors.New(resp.Errmsg)
```
Comment (tiancaiamao) [inline sessionctx/binloginfo/binloginfo.go]: log the error here, that information is important, and it may be overwrite by the next loop.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**